File: DSAcpp/Algorithm/Searching/ternarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

int ternarySearch(const vector<int> &arr, int a){
    int l=0,h=arr.size(),mid1,mid2;

    // divide the array into three sections
    while(h-l>5){
        mid1 = l+(h-l)/3;
        mid2 = h-(h-l)/3;

        if(a<arr[mid1]) h = mid1; // first part of three section
        else if(arr[mid2]<a) l = mid2; // last part of three section 
        else { // middle part of three section
            l = mid1;
            h = mid2;
        }
    }

    // return the index of the element
    for(int i=l;i<=h;i++) if(arr[i]==a) 
        return i;
    return -1;
}

int main(){
    vector<int> arr = {1,2,3,4,5,6,7,8,9,10,11,12,13,14};
    for(int i=0;i<15;i++)
        cout << ternarySearch(arr,i) << '\n';
    return 0;
}

File: DSApython/Algorithm/Searching/ternarySearch.py

def ternarySearch(arr,a):
    lw,hi = 0,len(arr)-1

    # divide the array into three sections
    while hi-lw>5:
        mid1 = lw+((hi-lw)//3)
        mid2 = hi-((hi-lw)//3)

        if a<arr[mid1]: hi = mid1 # first part of the three sections
        elif arr[mid2]<a: lw = mid2 # end part of the three sections
        else: # middle part of the three sections
            lw = mid1
            hi = mid2
    
    for i in range(lw,hi+1):
        if arr[i]==a: return i
    return -1

arr = [(i+1) for i in range(40)]

for a in range(10,30):
    ndx = ternarySearch(arr,a)
    print(f"seaching for {a}: index={"Not Found" if(ndx<0) else ndx}")

a = 0
ndx = ternarySearch(arr,a)
print(f"seaching for {a}: index={"Not Found" if(ndx<0) else ndx}")